src/util/buf.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use grpcio_sys::*;
+use std::ffi::c_void;
+use std::fmt::{self, Debug, Formatter};
+use std::io::{self, BufRead, Read};
+use std::mem::{self, ManuallyDrop, MaybeUninit};
+use std::ptr;
+use std::sync::atomic::{self, AtomicUsize, Ordering};
+
+/// A convenient rust wrapper for the type `grpc_slice`.
+#[repr(C)]
+pub struct GrpcSlice(grpc_slice);
+
+impl GrpcSlice {
+    /// Allocate a `GrpcSlice` with given length.
+    ///
+    /// It's unsafe to read from the slice as the content is uninitialized.
+    pub unsafe fn with_len(len: usize) -> GrpcSlice {
+        GrpcSlice(grpc_slice_malloc(len))
+    }
+
+    /// Get the length of the data.
+    pub fn len(&self) -> usize {
+        unsafe {
+            if !self.0.refcount.is_null() {
+                self.0.data.refcounted.length
+            } else {
+                self.0.data.inlined.length as usize
+            }
+        }
+    }
+
+    /// Returns a slice of inner buffer.
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe {
+            if !self.0.refcount.is_null() {
+                let start = self.0.data.refcounted.bytes;
+                let len = self.0.data.refcounted.length;
+                std::slice::from_raw_parts(start, len)
+            } else {
+                let len = self.0.data.inlined.length;
+                &self.0.data.inlined.bytes[..len as usize]
+            }
+        }
+    }
+
+    pub fn as_ptr(&self) -> *const u8 {
```

Comment:
This also seems to be unused and makes it much harder to ensure Rust's immutability invariant. Unless you have a use in mind, I would remove it.